### PR TITLE
refactor: replace external(v0) by abi(embed_v0)

### LINKED ISSUE #125: #[external(v0)] soon deprecated
We should replace it by `#[abi(embed_v0)]`

### PR DESCRIPTION
**Pull request by enitrat**
**Issue(s):** close #125 

### Description

- see title
- fixed wrong casing conventions in cheat sheet contracts (by the way, the relevancy of this section is not super obvious to me, but I guess the more the better)
- removed implicit interfaces (no longer possible - embedded interfaces can only be implementations of a starknet::interface trait)
### Checklist

- [x] **CI Verifier:** Run `cairo_programs_verifier` successfully
- [ ] **Contract Tests:** Added tests to cover the changes
- [ ] **Remix Link:** Added a link to open the contract in Remix
- [ ] **Deployed Contract (Optional):** If possible, include a Voyager link to a deployed contract on Goerli

### CODE REVIEW

File: src/ch00/interacting/interfaces-traits.md [hunk 1]
```diff
@@ -6,12 +6,9 @@ In cairo, to specify the interface you need to define a trait annotated with `#[
 
 When a function needs to access the contract state, it must have a `self` parameter of type `ContractState`. This implies that the corresponding function signature in the interface trait must also take a `TContractState` type as a parameter. It's important to note that every function in the contract interface must have this `self` parameter of type `TContractState`.
 
-You can use the `#[generate_trait]` attribute to implicitly generate the trait for a specific implementation block. This attribute automatically generates a trait with the same functions as the ones in the implemented block, replacing the `self` parameter with a generic `TContractState` parameter.
+You can use the `#[generate_trait]` attribute to implicitly generate the trait for a specific implementation block. This attribute automatically generates a trait with the same functions as the ones in the implemented block, replacing the `self` parameter with a generic `TContractState` parameter. However, this will only work for internal functions, and not for functions that are part of the contract's public interface - as embedding an impl inside the abi of a contract requires this impl to implement an interface.
```
Comment (julio4) [inline src/ch00/interacting/interfaces-traits.md:9]: See my comment in #125 (I think that `external(v0)` will be deprecated on the impl only)
Let's consider a contract with:
```rust
#[abi(per_item)]
#[generate_trait]
impl Contract of IContract {
    #[external(v0)]
    fn a(self: @ContractState) -> u32 {
        0
    }
}
```

Even with no explicit interface you can still declare your own dispatcher and use it, like this:
```rust
mod tests {
    #[starknet::interface]
    trait IGlobalExampleA<TContractState> {
        fn a(self: @TContractState) -> u32;
    }

    #[test]
    #[available_gas(2000000000)]
    fn test_a() {
        let (address0, _) = starknet::deploy_syscall(
            my_contracts::Contract::TEST_CLASS_HASH.try_into().unwrap(), 0, ArrayTrait::new().span(), false
        ).unwrap();

        let contract = IGlobalExampleADispatcher { contract_address: address0 };
        assert(contract.a() == 1, 'contract.(a) != 1');
    }
}
```

Comment (julio4) [reply src/ch00/interacting/interfaces-traits.md:9]: However I do agree that it's a best practice to always have one external functions block using `#[abi(embed_v0)]` than implement an explicitly defined interface, but it's not mandatory. Correct me if I'm wrong!

**Review by julio4 — COMMENTED** [review]
LGTM! Thank you for taking the time to correct these issues.

- About the Cheatsheet part: The idea was to have ultra concise example about common patterns (such as loop, match, type casting, ...) that new Cairo developers could refers to quickly remember the correct syntax.

File: src/ch00/interacting/interfaces-traits.md (hunk 1 above)
Comment (enitrat) [reply src/ch00/interacting/interfaces-traits.md:9]: No I think you're right!

**Comment by enitrat** [comment]
> About the Cheatsheet part: The idea was to have ultra concise example about common patterns (such as loop, match, type casting, ...) that new Cairo developers could refers to quickly remember the correct syntax.

As Cairo is not just a SC language it makes no sense to me why these would be presented inside contracts 🤔

**Comment by julio4** [comment]
> > About the Cheatsheet part: The idea was to have ultra concise example about common patterns (such as loop, match, type casting, ...) that new Cairo developers could refers to quickly remember the correct syntax.
> 
> As Cairo is not just a SC language it makes no sense to me why these would be presented inside contracts 🤔 It just introduces un-necessary boilerplate over what's actually important

I understand your concerns and mostly agree, it's a bit out of scope for `contracts`, but still in the philosophy of `by example`. I received some feedback that these quick references are helpful for beginners. Maybe we should move it to something similar in the Cairo Book instead?

**Comment by julio4** [comment]
LGTM! Thank you!

**Comment by enitrat** [comment]
I think it could stay there but just not wrapped inside a contract 🤔

**Comment by julio4** [comment]
> I think it could stay there but just not wrapped inside a contract 🤔

Hum I agree with you on that. It will be addressed soon.

**Pull request merged**